source/common/upstream/cluster_manager_impl.cc

```diff
@@ -343,6 +353,123 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
+bool ClusterManagerImpl::scheduleUpdate(const Cluster& cluster, uint32_t priority,
+                                        const HostVector& hosts_added,
+                                        const HostVector& hosts_removed) {
+  const auto& update_merge_window = cluster.info()->lbConfig().update_merge_window();
+  const auto timeout = DurationUtil::durationToMilliseconds(update_merge_window);
+  PendingUpdatesByPriorityMapPtr updates_by_prio;
+  PendingUpdatesPtr updates;
+
+  // Find pending updates for this cluster.
+  const auto updates_by_prio_it = updates_map_.find(cluster.info()->name());
+  if (updates_by_prio_it != updates_map_.end()) {
+    updates_by_prio = updates_by_prio_it->second;
+  } else {
+    updates_by_prio = std::make_shared<PendingUpdatesByPriorityMap>();
+    updates_map_[cluster.info()->name()] = updates_by_prio;
+  }
+
+  // Find pending updates for this priority.
+  const auto updates_it = updates_by_prio->find(priority);
+  if (updates_it != updates_by_prio->end()) {
+    updates = updates_it->second;
+  } else {
+    updates = std::make_shared<PendingUpdates>();
+    (*updates_by_prio)[priority] = updates;
+  }
+
+  // Has an update_merge_window gone by since the last update? If so, don't schedule
+  // the update so it can be applied immediately as long as there's no active timer
+  // (to avoid changing the order of updates).
+  const auto delta = std::chrono::steady_clock::now() - updates->last_updated;
+  const uint64_t delta_ms = std::chrono::duration_cast<std::chrono::milliseconds>(delta).count();
+  if (delta_ms > timeout && updates->timer == nullptr) {
+    updates->last_updated = std::chrono::steady_clock::now();
+    return false;
+  }
+
+  // Record the updates that should be applied when the timer fires.
+  // Handle the case when a pair of add/remove cancels out.
+  //
+  // A few notes:
+  // * if a host's health changed, we won't see that as an add or remove
+  //   so there's no need to check for health changes here. The fact that
+  //   we got called with no added/removed hosts implies a health or metadata
+  //   change, so we just forward the empty vectors along.
+  // * if while we are merging updates we see a host multiple times (e.g.:
+  //   add -> remove -> add), we deal with the cancellations and we always
+  //   propagate the latest version of that host.
+  // * we track hosts by their address since the Host ptr will most likely change
+  //   as it gets materialized out of an EDS change, and an address is considered
+  //   the unique constant identifier for a host.
+  for (const auto& host : hosts_added) {
+    const auto& addr = host->address()->asString();
+    // Is adding this host cancelling a previous remove?
+    if (updates->removed.count(addr) == 1) {
+      updates->removed.erase(addr);
+
+      // If it was previously added, store the latest one.
```

Comment:
If I have `[remove, add]` of a host, won't this check in `if` below skip the add? Can you add a test?